walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,79 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> attendees = new HashSet<String>(request.getAttendees());
+    List<TimeRange> initalRanges = new ArrayList<TimeRange>();
+    List<TimeRange> results = new ArrayList<TimeRange>();
+    final int minMeetingTime = 15;
+    
+    for(int start = 0; start <= TimeRange.END_OF_DAY; start += minMeetingTime) {
+        if(start + minMeetingTime > TimeRange.END_OF_DAY) {
+            initalRanges.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, false));
+        } else {
+            initalRanges.add(TimeRange.fromStartDuration(start, minMeetingTime));
+        }
+    }
+
+    for(Event currEvent : events) {
+        Set<String> currAttendees = currEvent.getAttendees();
+        for(String attendee : currAttendees) {
+            if(attendees.contains(attendee)) {
+                blockTimeRange(initalRanges, currEvent);
+                break;
+            }
+        }
+    }
+
+    TimeRange tempRange = new TimeRange();
+    int duration = (int) request.getDuration();
+
+// Group together the leftover time ranges for results
+    for(TimeRange range : initalRanges) {
+
+        //range is not schedulable and there was a previous valid range started
+        if(!range.isSchedulable() && tempRange.isValidRange()) {
+            if(tempRange.getDuration() >= duration) {
+            results.add(TimeRange.fromStartDuration(tempRange.getStart(), tempRange.getDuration()));
+            }
+            tempRange.setValidRange(false);
+            tempRange.setStart(0);
+            tempRange.setDuration(0);
```

Comment:
what if you phrased this as:

if (!range.isSchedulable()) {
   if(tempRange is valid) {
      save temprange
   }
   reset temp range
   continue;
}

here?

In fact - consider extracting the  if(temp range is valid) save temp range 

into its own function. I usually name these maybeAddToResult(tempRange, resultList) . You can then reuse this below! (see later comment)

All the rest of the code below can then assume that it will be adding to temp range and wont need the if (range is schedulable) check.